File: monitoring/Primes.java
import java.text.NumberFormat;
import java.util.LinkedList;

/**
 * Use link list to find all prime numbers from two to a billion.
 * Used to generate trafic to test monitoring of HPC resources.
 * 
 * @author Phillip Benoit
 *
 */
public class Primes {

	/**
	 * entry point
	 */
	public static void main(String[] args) {

		//used value because known number of primes
		int max = 1000000000;
		
		//number of divides for progress monitoring
		int slice = 100;
		
		//prevent making slice sizes that will not work
		if (max % slice != 0) {
			System.err.println("Slice did not evenly divide the max");
			System.exit(-1);
		}
		
		//percent formatting for monitoring
		NumberFormat percent = NumberFormat.getPercentInstance();
		percent.setMinimumFractionDigits(0);
		percent.setMaximumFractionDigits(1);

		//used for progress output
		int step = max/slice;

		//use linklist to store primes
		LinkedList<Integer> primes = new LinkedList<Integer>();

		//start with 2 as a known prime number
		primes.add(2);
		
		//number being tested for "primeness"
		//start x at the next number
		int x = 3;

		//used to print progress bar
		int y = step;

		//used to track progress
		for (; y <= max; y+=step) {

			//loop through numbers to test
			for (; x <= y; x+=2) {

				//use square root of possible prime as an upper limit to test
				int sqrt = (int) Math.sqrt(x);

				//step through list of prime numbers
				for (Integer test_number:primes) {

					//if the test number is larger than the square root, x is prime
					if (test_number > sqrt) {
						primes.add(x);
						break;
					}

					//if x is evenly divisible by a prime number, it is not prime
					if (x % test_number == 0) break;
				}
			}
			
			//monitoring output
			System.out.println(java.time.LocalDateTime.now() + " " + percent.
					format((double)x/max) + " Primes up to " + (x-1) + " found so "
							+ "far: " + primes.size());

		}

		//print number found
		System.out.println("Number of prime found from 2 - "
				+ "" + max + " = " + primes.size());

		//test against known list size
		// https://primes.utm.edu/howmany.html
		String correct = 50847534 == primes.size() ? "right number of primes "
				+ "found" : "wrong number of primes found";

		//print results
		System.out.println(correct);
	}

}

File: monitoring/primes.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

//define the structure for a linked list node
typedef struct node {
	long data;
	struct node* next;
}node;

//pointers for the head and tail of the list
static struct node *head = NULL, *tail = NULL;

//count starts at 2 (2 and 3 are considered to be part of the list)
long counter = 2;

//prime numbers to a billion
#define LAST_NUMBER 1000000000

//slices for progress
#define SLICE 100

//adds a link to the linked list
void addNode(long n) {
	
	//allocate space for the node
	struct node* newNode = (struct node*)malloc(sizeof(struct node));
	
	//if malloc fails to allocate, print a helpful message
	if (newNode == NULL) printf("Stopped at %d. size of node: %d\n", counter,
		sizeof(struct node));
		
	//set new prime
	newNode->data = n;
	
	//point to new prime
	tail->next = newNode;
	
	//set new prime as the tail of the list
	tail = newNode;
	
	//increment counter
	counter++;
}

//entry point
int main() {

	//reject bad math that prevents proper counting
	if (LAST_NUMBER % SLICE != 0) {
		printf("Slice did not evenly divide the max\n");
		return -1;
	}

	//define step for monitoring loop
	long step = LAST_NUMBER / SLICE;

	//add 3 as first node of the list
	head = (struct node*)malloc(sizeof(struct node));
	head->data = 3;
	tail = head;

	//number being tested for primeness
	//start counting from 5
	long x = 5;

	//outer loop counter for monitoring
	long y = step;

	//cursor for stepping through the linked list
	struct node* cursor;

	//Holds the square root of the number being tested for "primeness".
	//Only prime numbers less than or equal
	//to this value are possible factors.
	long squarert;

	//variables that hold time info for monitoring
	time_t rawtime;
	struct tm * timeinfo;

	//outer monitoring loop
	for (; y <= LAST_NUMBER; y += step) {
		
		//step through numbers to test for "primeness"
		for (; x <= y; x += 2) {
			
			//find largest possible factor
			squarert = (long)sqrtl(x);
			
			//start at the beginning of the primes linked list
			cursor = head;
			
			//step through the linked list while primes do not evenly divide
			//the number being tested
			while (x % cursor->data != 0) {
			
				//break loop if a prime number is found
				if (cursor->data > squarert) {
					addNode(x);
					break;
				}
				
				//steps through the list
				cursor = cursor->next;
			}
		}

		//get time info
		time(&rawtime);
		timeinfo = localtime(&rawtime);
		
		//display progress
		printf("%.1f%% %d primes found ", ((double)x / LAST_NUMBER) * 100, counter);
		printf(asctime(timeinfo));
	}

	//see if results are correct
	if (counter == 50847534) printf("Correct number of primes found.\n");
	else printf("Incorrect number of primes found.\n");

	return 0;
}
